packages/astro/src/cli/index.ts

```diff
@@ -132,15 +132,44 @@ async function runCommand(cmd: string, flags: yargs.Arguments) {
 		}
 	}
 
-	const { astroConfig, userConfig } = await openConfig({ cwd: root, flags, cmd });
+	let { astroConfig, userConfig, userConfigPath } = await openConfig({ cwd: root, flags, cmd });
 	telemetry.record(event.eventCliSession(cmd, userConfig, flags));
 
 	// Common CLI Commands:
 	// These commands run normally. All commands are assumed to have been handled
 	// by the end of this switch statement.
 	switch (cmd) {
 		case 'dev': {
-			await devServer(astroConfig, { logging, telemetry });
+			async function startDevServer() {
+				const { watcher, stop } = await devServer(
+					astroConfig,
+					{ logging, telemetry },
+				);
+
+				watcher.on('change', logRestartServerOnConfigChange);
+				watcher.on('unlink', logRestartServerOnConfigChange);
+				function logRestartServerOnConfigChange(changedFile: string) {
+					if (userConfigPath === changedFile) {
+						warn(logging, 'astro', 'Astro config updated. Restart server to see changes!');
+					}
+				}
+
+				watcher.on('add', async function restartServerOnNewConfigFile(addedFile: string) {
+					// if there was not a config before, attempt to resolve
+					if (!userConfigPath && addedFile.includes('astro.config')) {
```

Comment:
I included that fuzzy `addedFile.includes('astro.config')` to avoid the work-intensive `openConfig` on _most_ file changes.